subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/PrecompileGroovyScriptsTask.java

```diff
@@ -0,0 +1,299 @@
+        return pluginAdapterClassesOutputDir.getAsFile();
+    }
+
+    @TaskAction
+    void compileScripts() {
+        FileCollection compileClasspath = classpath.plus(project.files(javaSourceDependencyClasses, groovySourceDependencyClasses));
+        ClassLoader compileClassLoader = new URLClassLoader(DefaultClassPath.of(compileClasspath).getAsURLArray(), classLoaderScope.getLocalClassLoader());
+
+        for (PrecompiledGroovyScript scriptPlugin : scriptPlugins) {
+            CompiledScript<PluginsAwareScript, ?> pluginsBlock = compilePluginsBlock(scriptPlugin, compileClassLoader);
+            PluginRequests pluginRequests = getValidPluginRequests(pluginsBlock, scriptPlugin, compileClassLoader);
+            CompiledScript<? extends BasicScript, ?> buildScript = compileBuildScript(scriptPlugin, compileClassLoader);
+            generateScriptPluginAdapter(scriptPlugin, pluginRequests, buildScript);
+        }
+
+        fileSystemOperations.copy(copySpec -> {
+            copySpec.from(intermediatePluginClassesDir.get().getAsFileTree().getFiles());
+            copySpec.into(precompiledGroovyScriptsOutputDir);
+        });
+    }
+
+    private PluginRequests getValidPluginRequests(CompiledScript<PluginsAwareScript, ?> pluginsBlock, PrecompiledGroovyScript scriptPlugin, ClassLoader compileClassLoader) {
+        if (!pluginsBlock.getRunDoesSomething()) {
+            return PluginRequests.EMPTY;
+        }
+        PluginRequests pluginRequests = extractPluginRequests(pluginsBlock, scriptPlugin, compileClassLoader);
+        Set<String> validationErrors = new HashSet<>();
+        for (PluginRequest pluginRequest : pluginRequests) {
+            if (pluginRequest.getVersion() != null) {
+                validationErrors.add(String.format("Invalid plugin request %s. " +
+                        "Plugin requests from precompiled scripts must not include a version number. " +
+                        "Please remove the version from the offending request and make sure the module containing the " +
+                        "requested plugin '%s' is an implementation dependency of %s",
+                    pluginRequest, pluginRequest.getId(), project));
+            }
+        }
+        if (!validationErrors.isEmpty()) {
+            throw new LocationAwareException(new IllegalArgumentException(String.join("\n", validationErrors)),
+                scriptPlugin.getSource().getResource().getLocation().getDisplayName(),
+                pluginRequests.iterator().next().getLineNumber());
+        }
+        return pluginRequests;
+    }
+
+    private PluginRequests extractPluginRequests(CompiledScript<PluginsAwareScript, ?> pluginsBlock, PrecompiledGroovyScript scriptPlugin, ClassLoader compileClassLoader) {
+        ScriptRunner<PluginsAwareScript, ?> runner = scriptRunnerFactory.create(pluginsBlock, scriptPlugin.getSource(), compileClassLoader);
+
+        Project target = ProjectBuilder.builder().withParent(project).build();
```

Comment:
I think we should avoid trying to use `ProjectBuilder` since it's intended to be a test-only thing.  I think we want to do something similar to https://github.com/gradle/gradle/blob/79dd695beeb3c272bb9f4053104dee98647ed8db/subprojects/core/src/main/java/org/gradle/configuration/DefaultScriptPluginFactory.java#L107-L113

Reply:
The thing is that we don't want to apply the plugins to the compiling project here - we rather want to execute the script against `something` in order to collect the plugin requests and validate them.
Turns out that we don't need to pass a project here and any non-null object passed does the trick. Like `runner.run("dummy", serviceRegistry);`. But I do not like this.